apm-agent-plugins/apm-httpclient11-plugin/src/main/java/co/elastic/apm/agent/httpclient/HttpClientAsyncInstrumentation.java

```diff
@@ -0,0 +1,155 @@
+
+    @VisibleForAdvice
+    public static final WeakConcurrentMap<CompletableFuture<?>, Span> handlerSpanMap = WeakMapSupplier.createMap();
+
+    @VisibleForAdvice
+    public static class HttpClient11Advice {
+
+        @VisibleForAdvice
+        public final static GlobalThreadLocal<Span> spanTls = GlobalThreadLocal.get(HttpClient11Advice.class, "spanTls");
+
+        @Nullable
+        @AssignTo.Argument(value = 0, typing = Assigner.Typing.DYNAMIC)
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object onBeforeExecute(@Advice.Argument(value = 0) HttpRequest httpRequest) {
+            if (tracer.getActive() == null) {
+                return httpRequest;
+            }
+            final AbstractSpan<?> parent = tracer.getActive();
+            URI uri = httpRequest.uri();
+            Span span = co.elastic.apm.agent.http.client.HttpClientHelper.startHttpClientSpan(parent, httpRequest.method(), uri.toString(), uri.getScheme(),
+                HttpClientHelper.computeHostName(uri.getHost()), uri.getPort());
+            if (span != null) {
+                spanTls.set(span);
+                span.activate();
+                HttpRequest.Builder builder = HttpRequest.newBuilder(httpRequest.uri())
+                    .method(httpRequest.method(), httpRequest.bodyPublisher().orElse(HttpRequest.BodyPublishers.noBody()))
+                    .expectContinue(httpRequest.expectContinue());
+                if (httpRequest.timeout().isPresent()) {
+                    builder = builder.timeout(httpRequest.timeout().get());
+                }
+                if (httpRequest.version().isPresent()) {
+                    builder = builder.version(httpRequest.version().get());
+                }
+                for (String header : httpRequest.headers().map().keySet()) {
+                    builder.header(header, httpRequest.headers().firstValue(header).orElse(null));
+                }
+                return builder.build();
+            }
+            return httpRequest;
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void onAfterExecute(@Advice.Return @Nullable HttpResponse response,
+                                          @Advice.Thrown @Nullable Throwable t) {
+            final Span span = spanTls.getAndRemove();
+            if (span != null) {
+                try {
+                    if (response != null) {
+                        int statusCode = response.statusCode();
+                        span.getContext().getHttp().withStatusCode(statusCode);
+                    }
+                    span.captureException(t);
+                } finally {
+                    span.deactivate().end();
+                }
+            }
+        }
+    }
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return named("jdk.internal.net.http.HttpClientImpl");
```

Comment:
don't depend on the exact internal name. instead, use `hasSuperType(named("java.net.http.HttpClient"))` and add a pre matcher with `nameContains("HttpClient")`

Reply:
fixed